### PR TITLE
added cookie checking

### PR DESCRIPTION
**Pull request by angsm**
Added cookie checking and redirection to /auth on server.js.

### CODE REVIEW

File: projects/gnomad/config/webpack.config.client.js [hunk 1]
```diff
@@ -13,6 +13,7 @@ if (process.env.NODE_ENV === 'production' && !gaTrackingId) {
 
 const config = {
   devServer: {
+    writeToDisk: true,
```
Comment (mhebrard) [inline projects/gnomad/config/webpack.config.client.js]:
```suggestion
```
We should not need to change devServer config, at it do not impact prod

File: projects/gnomad/src/server/server.js [hunk 2]
```diff
@@ -2,12 +2,21 @@ import path from 'path'
 
 import compression from 'compression'
 import express from 'express'
+import cookieParser from 'cookie-parser'
+
+const winston = require('winston');
```
Comment (mhebrard) [inline projects/gnomad/src/server/server.js]:
```suggestion
import winston from winston
```
Prefer import syntax

File: projects/gnomad/src/server/server.js [hunk 3]
```diff
@@ -2,12 +2,21 @@ import path from 'path'
...
+const logger = winston.createLogger({
+  transports: [
+    new winston.transports.Console(),
+    new winston.transports.File({ filename: '/opt/web/gnomadjs/projects/gnomad/src/server/access.log' })
```
Comment (mhebrard) [inline projects/gnomad/src/server/server.js]:
```suggestion
    new winston.transports.File({ filename: path.resolve(__dirname, 'access.log') })
```
Full path should not be hard coded... I am not sure but we should be able to use  `path.resolve(__dirname, ...)`

File: projects/gnomad/src/server/server.js [hunk 4]
```diff
@@ -29,8 +38,56 @@ if (JSON.parse(process.env.ENABLE_HTTPS_REDIRECT || 'false')) {
...
+//  let decoded = jwtDecode(chorus_accessToken)
+//  var current_time = (new Date().getTime() / 1000)
+  
+//  if (current_time > decoded.exp){
+//    console.log('EXPIRED')
+//    res.redirect(`127.0.0.1/auth`)  
+//  }
+//  next()
+//}
+//app.use(myLogger)
```
Comment (mhebrard) [inline projects/gnomad/src/server/server.js]: If `myLogger` is not used, please remove the code

File: projects/gnomad/src/server/server.js [hunk 5]
```diff
@@ -29,8 +38,56 @@ if (JSON.parse(process.env.ENABLE_HTTPS_REDIRECT || 'false')) {
...
+
+app.use((request, response, next) => {
+  // get cookies
+  var isRedirect = true
+  var jwtDecode = require('jwt-decode')
```
Comment (mhebrard) [inline projects/gnomad/src/server/server.js]:
```suggestion
  import jwtDecode from 'jwt-decode'
```
Prefer an import syntax on top of the code

File: projects/gnomad/src/server/server.js [hunk 6]
```diff
@@ -29,8 +38,56 @@ if (JSON.parse(process.env.ENABLE_HTTPS_REDIRECT || 'false')) {
...
+    if ((new Date().getTime() / 1000) < decoded.exp){ //check if not expired
+      console.log('isVALID: ', true)
+      isRedirect = false
+    }
+  }
+
+  if (isRedirect === true) {
+    console.log('isVALID: ', false)
+    response.clearCookie('accessToken', {path: '/'})
+    response.redirect(302,'http://46.137.240.42/auth')
```
Comment (mhebrard) [inline projects/gnomad/src/server/server.js]:
```suggestion
    response.redirect(302,'http://localhost/auth')
```
the IP should not be hard codded... does localhost works ?

File: projects/gnomad/src/server/server.js.bak [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+const publicDir = path.resolve(__dirname, 'public')
+app.use(express.static(publicDir))
+app.get('*', (request, response) => {
+  response.sendFile(path.join(publicDir, 'index.html'))
+})
+
+app.listen(port, () => {
+  // eslint-disable-next-line no-console
+  console.log(`Listening on ${port}`)
+})
```
Comment (mhebrard) [inline projects/gnomad/src/server/server.js.bak]: This file should be ignored ( add an entry for .bak in `.gitignore` )

File: projects/gnomad/package.json [hunk 8]
```diff
@@ -15,7 +15,12 @@
   ],
   "dependencies": {
     "compression": "^1.7.4",
-    "express": "^4.16.4"
+    "cookie-parser": "^1.4.4",
+    "express": "^4.16.4",
+    "express-redirect-loop": "0.0.8",
+    "jwt-decode": "^2.2.0",
+    "universal-cookie": "^4.0.3",
+    "winston": "^3.2.1"
```
Comment (mhebrard) [inline projects/gnomad/package.json]: `express-redirect-loop` and `universal-cookie` are not used. they should be removed

Comment (angsm) [reply projects/gnomad/package.json]: Used universal cookie in the upcoming logout portion. will remove express-redirect-loop. That installation actually error out, that is weird.

File: projects/gnomad/src/server/server.js.bak (hunk 7 above)
Comment (angsm) [reply projects/gnomad/src/server/server.js.bak]: yeah, will do.

File: projects/gnomad/src/client/NavBar.js [hunk 9]
```diff
@@ -152,6 +152,8 @@ class NavBar extends Component {
                 FAQ
               </Link>
             </li>
+
+
```
Comment (mhebrard) [inline projects/gnomad/src/client/NavBar.js]: Try not to have diff on this file

File: projects/gnomad/src/client/routes.js [hunk 10]
```diff
@@ -21,6 +21,9 @@ import RegionPageContainer from './RegionPage/RegionPageContainer'
 import TranscriptPageContainer from './TranscriptPage/TranscriptPageContainer'
 import VariantPage from './VariantPage/VariantPage'
 
+import Logout from './Logout'
```
Comment (mhebrard) [inline projects/gnomad/src/client/routes.js]: To keep things clean I would do 
```suggestion
import LogoutPage from './ChorusPages/Logout'
```

File: projects/gnomad/src/client/routes.js [hunk 11]
```diff
@@ -21,6 +21,9 @@ import RegionPageContainer from './RegionPage/RegionPageContainer'
...
+
+
```
Comment (mhebrard) [inline projects/gnomad/src/client/routes.js]: no need new lines here

File: projects/gnomad/src/server/server.js [hunk 12]
```diff
@@ -29,8 +40,39 @@ if (JSON.parse(process.env.ENABLE_HTTPS_REDIRECT || 'false')) {
   })
 }
 
+app.use((request, response, next) => {
+  // get cookies
+  var isRedirect = true
+  var cookieVal = request.cookies['auth_accessToken']
+  var cookieVal = request.cookies['auth_idToken']
+  console.log('Cookie check', 'YES')
```
Comment (mhebrard) [inline projects/gnomad/src/server/server.js]: Delete console.log
```suggestion
```

File: projects/gnomad/src/server/server.js [hunk 13]
```diff
@@ -29,8 +40,39 @@ if (JSON.parse(process.env.ENABLE_HTTPS_REDIRECT || 'false')) {
...
+
+  if (typeof cookieVal !== 'undefined' && cookieVal !== null ) {
+    let decoded = jwtDecode(cookieVal)
+
+    if ((new Date().getTime() / 1000) < decoded.exp){ //check if not expired
+      console.log('isVALID: ', true)
```
Comment (mhebrard) [inline projects/gnomad/src/server/server.js]:
```suggestion
      logger.verbose('Cookie is valid')
```

File: projects/gnomad/src/server/server.js [hunk 14]
```diff
@@ -29,8 +40,39 @@ if (JSON.parse(process.env.ENABLE_HTTPS_REDIRECT || 'false')) {
...
+      isRedirect = false
+    }
+  }
+
+  if (isRedirect === true) { //redirect to authentication page
+    console.log('isVALID: ', false)
```
Comment (mhebrard) [inline projects/gnomad/src/server/server.js]:
```suggestion
    logger.verbose('Redirecting')
```

File: projects/gnomad/src/server/server.js [hunk 15]
```diff
@@ -29,8 +40,39 @@ if (JSON.parse(process.env.ENABLE_HTTPS_REDIRECT || 'false')) {
...
+    response.clearCookie('auth_accessToken', {path: '/'}) //clear cookies
+    response.clearCookie('auth_idToken', {path: '/'}) //clear cookies
+    response.redirect(302, `http://${request.get('host')}/auth`)
+  }else {
+    console.log('STAY', 'YES')
```
Comment (mhebrard) [inline projects/gnomad/src/server/server.js]: Delete console.log
```suggestion
```

File: projects/gnomad/src/client/Logout.js [hunk 16]
```diff
@@ -0,0 +1,20 @@
+import React from 'react'
+import { withRouter, Link, Redirect } from 'react-router-dom'
+import Cookies from 'universal-cookie'
+
+class Logout extends React.Component {
```
Comment (mhebrard) [inline projects/gnomad/src/client/Logout.js]: I would move this file to `ChorusPages` and rename it `LogoutPage.js`
rename the class and the export to `LogoutPage`

File: projects/gnomad/src/server/server.js (hunk 6 above)
Comment (angsm) [reply projects/gnomad/src/server/server.js]: Updated in next commit

File: projects/gnomad/src/server/server.js.bak (hunk 7 above)
Comment (angsm) [reply projects/gnomad/src/server/server.js.bak]: Updated in next commit

File: projects/gnomad/src/client/Logout.js (hunk 16 above)
Comment (angsm) [reply projects/gnomad/src/client/Logout.js]: Updated in next commit

File: projects/gnomad/src/client/routes.js (hunk 10 above)
Comment (angsm) [reply projects/gnomad/src/client/routes.js]: Done on next commit

**Pull request merged**